### DISCUSSION TITLE
Using `vscode.workspace.fs.readFile` to access builtin extension data from user extension

### DISCUSSION THREAD

**Discussion by zm-cttae** [microsoft/vscode-discussions#553] [Extension Development QnA] [answered]
On electron (desktop platform), VS Code allows user extensions to use fs.readFile in order to access assets from the core application itself. E.g. builtin extension files. However, this doesn't work on web - where it appears that the extension host worker and core app are hosted on their own domains each with their own separate ersatz filesystem via `vscode.workspace.fs`.

For fs.readFile on web, is there a theoretical capacity to access the core app from user extensions? A good application is reading the source of a language configuration JSON or Textmate grammar

@jrieken @alexdima @sandy081

**Comment by jrieken** [comment]
`fs.readFile` is specific to Nodejs and the web worker extension host isn't a node process. To ensure web-readiness avoid the use nodejs API but use the VS Code API, e.g `vscode.workspace.fs.readFile`. This works for both extension host kinds - web and desktop.

**Reply by zm-cttae** [reply]
I'm aware - I want to do cross-extension file reads in the web runtime, and read from `$appRoot/resources/app/extensions`.

Whoops I mentioned the wrong API & method in the title 😄 fixed

**Reply by jrieken** [reply, marked as answer]
You should be successful when using [`Extension.extensionUri`](https://github.com/microsoft/vscode/blob/99d65cd47c9ce741f7e703b48df8e1251ed2edeb/src/vscode-dts/vscode.d.ts#L7025)

**Reply by zm-cttae** [reply]
Huh.. didn't try that. That's simple and genius! Uri has server property so it can resolve these concerns :) can mark as answer now!

**Reply by zm-cttae** [reply]
Confirmed this works and is getting me the data I need :)
